### DISCUSSION TITLE
Problem with converting .ARW to exr

### DISCUSSION THREAD

**Discussion by hereJonesy** [ImageMagick/ImageMagick#2958] [Digital Image Processing]
Hi,
I'm trying to convert .ARW files to linear exr's and am getting some colour issues - the images are coming out all pink. 
Would somebody be able to advise on what other options I might need to use?
Thanks

**Comment by fmw42** [comment]
Please provide your IM version and platform/OS. Also provide an example input and output image and your exact command line.

**Reply by hereJonesy** [reply]
IM version: 7.0.10-46 Q16 x86_64 2020-12-07
OS: CentOS Linux
Command line: magick convert [image].ARW [image].exr
Unfortunately I'm unable to upload the input/output image for security reasons. 

I'm sure there are some colourspace commands I'm missing, but unfortunately it's a bit out of my area of expertise!

**Comment by snibgo** [comment]
Diagnosis without the images is difficult. It might be a white balance problem, so perhaps "-define dng:use-camera-wb=true" before the input file will help. See also other DNG defines at https://imagemagick.org/script/defines.php.

EDIT to add: EXR brings in other variables. For testing, try writing to TIFF instead of EXR. Does that give the expected result?

**Comment by hereJonesy** [comment]
Thanks for you help.
The whitebalance command is giving the same incorrect result. I also tried a TIFF and it looks the same :(

**Comment by fmw42** [comment]
What is your version of OpenEXR delegate library? Is it current?  Can you reproduce by taking some non-proprietary ARW image and convert that to EXR? If so, then post that input and output non-proprietary pair of images and your exact command line. Note that with IM 7, use magick, not magick convert.

**Comment by hereJonesy** [comment]
I have uploaded the .ARW to my google drive:
https://drive.google.com/file/d/1maBTHX95k02dMR_7iZAywR48hphyzXOt/view?usp=sharing

Here is an example of the result I'm getting from the conversion:
https://drive.google.com/file/d/1zVub4scfBsfXIuZ8dWnmxDHEa73RhkR6/view?usp=sharing

The TIF file above was produced using the following command:
magick -define dng:use-camera-wb=true -define dng:output-color=1 S03_LD_20201008_ld_001.ARW test05.TIF

However I've tried all manner of combinations, with/without the colourspace commands and they all come out the same.

The above example is trying to create an sRGB tiff but the end goal is to try and get an ACES exr (which I've also tried with the same pinkish result).

Thanks

**Comment by fmw42** [comment]
Please post your images so that one does not need an account at gmail or need to ask for permission to download the files.

**Reply by hereJonesy** [reply]
Apologies - I hadn't specified the sharing settings correctly. 
These links should download without requiring access or needing a gmail account:

https://drive.google.com/file/d/17GFqUGhovQA9tD2HfIQBpdwMtiG8mT3b/view?usp=sharing
https://drive.google.com/file/d/19EmVEUae8bt7-Fdwme65oPrjxxdrR22q/view?usp=sharing

Thanks again

**Comment by snibgo** [comment]
Google won't download your second file, the TIFF, possibly because it is so large (244MB). Can you upload a smaller version?

Developing your ARW with either ImageMagick (using libraw) or dcraw, most gray patches have a slight bias to red. For example, the middle gray is about RGB(67%,63%,58%). Is that what you mean by "coming out all pink"?

A slight tweak to the colour balance will fix that, eg multiply red by 63/67 and blue by 63/58.

**Reply by hereJonesy** [reply]
Here's a smaller version of the tiff!
https://drive.google.com/file/d/1Fzk5qI-tSjlNSmiKuNbgANjOF8n1GwMa/view?usp=sharing

I will have a go at those grade tweaks and let you know. Assuming I can apply that within my command line process?

**Comment by snibgo** [comment]
For comparison, here is what I get with IM v7.0.8-64 and v7.0.10-45, using libraw:
```
magick S03_LD_20201008_ld_001.ARW -resize 800x800 x.jpg
```
![x](https://user-images.githubusercontent.com/33812368/102119986-d6a30d00-3e39-11eb-8c19-66a7da39cc13.jpg)

There is a slight red shift, but no big problem. "-define dng:use-camera-wb=true -define dng:output-color=1" is slightly different, but nothing major.

Your scaled version shows a much stronger shift to red than I get, eg the middle gray patch (third from lightest) is about sRGB(50%,37%,40%). We can correct that:
```
magick S03_LD_20201008_ld_001_scaled.TIF -channel R -evaluate Multiply %[fx:37/50] -channel B -evaluate Multiply %[fx:37/40] +channel x.tiff
```
This corrects that gray patch, making it about gray(37%). But lighter patches then have green bias, and darker patches have a red bias. This suggests a colorspace problem, rather than a simpler white-balance problem.

Please paste here the text output from:
```
magick -version
```

Did you build IM yourself, or download a pre-built binary?

**Reply by hereJonesy** [reply]
Here is the output:

```
Version: ImageMagick 7.0.10-46 Q16 x86_64 2020-12-07 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC OpenMP(3.1) 
Delegates (built-in): bzlib djvu fontconfig freetype gvc jbig jng jpeg lcms lqr lzma openexr png raqm tiff x xml zlib
```

I believe our systems guys installed a pre-built binary. The first one in this Unix pre-built list:
https://imagemagick.org/script/download.php

I tried converting the same images using ufraw and it's giving the same result. Could it be that IM is defaulting to ufraw and we need to build from scratch to read RAW using libraw?

**Comment by snibgo** [comment]
You don't have libraw (it would show as "raw" in delegates), so your IM will use either ufraw or dcraw. Maybe "magick -verbose S03_LD_20201008_ld_001.ARW x.tif" will say. Perhaps ufraw is misreading your ARW file, and an update might fix it.

You might also look at your delegates.xml. Does that use dcraw or ufraw? Are the options reasonable?

If you have dcraw, try using that. It may be in your IM installation directory.

The best long-term solution may be to re-build IM with libraw.

**Reply by hereJonesy** [reply]
Hi, 

Thanks for all the help so far. 
Our systems team have re-built IM using libraw. 

Now when running magick -version we get:

```
Version: ImageMagick 7.0.10-50 Q16 x86_64 2020-12-18 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(3.1) 
Delegates (built-in): bzlib djvu fftw fontconfig freetype jng jpeg lcms lzma pangocairo png raw tiff x xml zlib
```

If I run:
`magick S03_LD_20201008_ld_001.ARW -resize 800x800 test06.jpg`

it now comes out looking quite green 
![test06](https://user-images.githubusercontent.com/75756927/103558047-fadb9200-4eab-11eb-8e34-973bf5dced46.jpg)

However if I run:
`magick -define dng:use-camera-wb=true S03_LD_20201008_ld_001.ARW -resize 800x800 test07.jpg`

it looks correct!
![test07](https://user-images.githubusercontent.com/75756927/103558148-1fd00500-4eac-11eb-806c-a15c0f615cc8.jpg)

Ultimately I want to convert to ACEScg exrs. So I tried running the following (unsure if this is the correct command):
`magick -verbose -define dng:use-camera-wb=true -define dng:output-color=6 S03_LD_20201008_ld_001.ARW test06.exr`

and I get the following:
```
app_wrapper: ==== starting app ====
S03_LD_20201008_ld_001.ARW ARW 5320x7968 8000x5320+0+0 16-bit sRGB 81.5322MiB 9.500u 0:01.865
S03_LD_20201008_ld_001.ARW=>test06.exr ARW 5320x7968 8000x5320+0+0 16-bit sRGB 81.5322MiB 0.000u 0:00.000
magick: UnableToOpenConfigureFile `delegates.xml' @ warning/configure.c/GetConfigureOptions/702.
magick: NoEncodeDelegateForThisImageFormat `ARW' @ error/constitute.c/WriteImage/1252.
```

Any help on where to go from here would be very appreciated! Thanks.

**Comment by snibgo** [comment]
> UnableToOpenConfigureFile `delegates.xml'

That is bad news. Something is wrong with your installation.

Try the command with "-debug all". In the very long text output, IM will say what directories are being searched for that file.

Until that problem is resolved, I suggest you test with very simple commands, such as:
```
magick rose: r.png
magick r.png r.jpg
magick r.png r.exr
```

**Reply by hereJonesy** [reply]
Thanks, I think we've fixed the delegate error.

Just so we know how to properly test it - would you be able to give me the correct command for converting a .ARW file to an ACEScg .exr and ensuring that it uses libraw?

**Comment by snibgo** [comment]
> ... would you be able to give me the correct command for converting a .ARW file to an ACEScg .exr and ensuring that it uses libraw?

Short answer: no.

Long answer:

I don't use EXR files and have nothing other than ImageMagick to check they are okay. When IM uses libraw to read raw files, it doesn't embed an ICC profile. So we don't know if the result from reading the ARW is linear or gamma 1.8 or something else.

In addition, IM doesn't embed ICC profiles into EXR files; I don't know if this is limitation of EXR or ICC.

However, gamma 1.8 seems to roughly work. All code is Windows BAT syntax. This tries to make an EXR, encoded as ACEScg (which is linear).
```
magick ^
  -define dng:output-color=6 -define dng:use-camera-wb=true S03_LD_20201008_ld_001.ARW ^
  -profile ACES-elle-V4-g18.icc ^
  -profile ACEScg-elle-V4-g10.icc ^
  x.exr
```
Profiles with "elle" in the name are available from https://ninedegreesbelow.com/photography/lcms-make-icc-profiles.html

To check, we can convert x.exr to sRGB:
```
magick ^
  x.exr ^
  -profile ACEScg-elle-V4-g10.icc ^
  -profile sRGB.icc ^
  x.tiff
```
x.tiff look roughly okay.

Personally, I don't use ImageMagick to read camera raw files because (a) it doesn't provide the full range of options available in dcraw or dcraw_emu and (b) the result image doesn't contain an embedded ICC profile.

So, repeating the above using dcraw_emu to read the ARW and make a tiff, which we then process with IM:
```
dcraw_emu -v -o 6 -w -T -6 S03_LD_20201008_ld_001.ARW

magick ^
  S03_LD_20201008_ld_001.ARW.tiff ^
  -verbose info: |grep -i icc
```
The text says:
```
    icc:copyright: auto-generated by dcraw
    icc:description: ACES
    Profile-icc: 476 bytes
```
... so dcraw_emu has embedded an ICC profile in the TIFF.

Convert that TIFF to ACEScg (which is linear):
```
magick ^
  S03_LD_20201008_ld_001.ARW.tiff ^
  -profile ACEScg-elle-V4-g10.icc ^
  x2.exr
```

Check, by converting the EXR to sRGB:
```
magick ^
  x2.exr ^
  -profile ACEScg-elle-V4-g10.icc ^
  -profile sRGB.icc ^
  x2.tiff
```
x.tiff and x2.tiff are similar but not identical. I suspect, but can't be sure, that x2.tiff is more correct.
